src/hotspot/os/linux/os_linux.cpp

```diff
@@ -238,12 +236,53 @@ julong os::Linux::available_memory() {
     }
   }
 
-  sysinfo(&si);
-  avail_mem = (julong)si.freeram * si.mem_unit;
+  FILE *fp = os::fopen("/proc/meminfo", "r");
+  if (fp) {
+    char buf[80];
+    do {
+      julong mem_available;
+      if (fscanf(fp, "MemAvailable: " JULONG_FORMAT " kB", &mem_available) == 1) {
+        avail_mem = mem_available;
+        break;
+      }
+    } while (fgets(buf, sizeof(buf), fp) != nullptr);
+    fclose(fp);
+  }
+  if (avail_mem == 0UL) {
+    avail_mem = free_memory();
+  }
   log_trace(os)("available memory: " JULONG_FORMAT, avail_mem);
   return avail_mem;
 }
 
+julong os::free_memory() {
+  return Linux::free_memory();
+}
+
+julong os::Linux::free_memory() {
+  // values in struct sysinfo are "unsigned long"
+  struct sysinfo si;
+  julong free_mem;
+
+  if (OSContainer::is_containerized()) {
+    jlong mem_limit = OSContainer::memory_limit_in_bytes();
+    jlong mem_usage;
+    if (mem_limit > 0 && (mem_usage = OSContainer::memory_usage_in_bytes()) < 1) {
+      log_debug(os, container)("container memory usage failed: " JLONG_FORMAT ", using host value", mem_usage);
+    }
+    if (mem_limit > 0 && mem_usage > 0) {
+      free_mem = mem_limit > mem_usage ? (julong)mem_limit - (julong)mem_usage : 0;
+      log_trace(os)("free container memory: " JULONG_FORMAT, free_mem);
+      return free_mem;
+    }
+  }
```

Comment:
We now have this code twice in `os_linux.cpp`. Could we parameterize this and extract as a function so that we avoid the duplication from lines 226-236?